bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,237 @@
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Path;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+public class FileRecoveryLog implements RecoveryLog {
+    private static final Logger log = LoggerFactory.getLogger(FileRecoveryLog.class);
+    private String baseFileName;
+    private Path recoveryLogDir;
+    private int checkpointInterval;
+    private final boolean deleteOldLogs;
+    private int numOfPutsSinceLastCheckpoint;
+    private File file;
+    private FileChannel appendChannel = null;
+    private Map<String, TransactionLogRecord> existingLogs;
+    private static final PrintStream stderr = System.err;
+
+    /**
+     * Initializes a new FileRecoveryLog instance with the given base file name.
+     *
+     * @param baseFileName  The base name for the recovery log files.
+     * @param checkpointInterval The interval at which to write checkpoints to the log file.
+     * @param recoveryLogDir The directory to store the recovery log files in.
+     * @param deleteOldLogs Whether to delete old log files when creating a new one.
+     */
+    public FileRecoveryLog(String baseFileName, int checkpointInterval, Path recoveryLogDir, boolean deleteOldLogs) {
+        this.baseFileName = baseFileName;
+        this.recoveryLogDir = recoveryLogDir;
+        this.deleteOldLogs = deleteOldLogs;
+        this.checkpointInterval = checkpointInterval;
+        this.existingLogs = new HashMap<>();
+        this.file = createNextVersion();
+        this.numOfPutsSinceLastCheckpoint = 0;
+    }
+
+    /**
+     * Creates the next version of the recovery log file, cleaning up the previous one.
+     *
+     * @return The newly created log file.
+     */
+    private File createNextVersion() {
+        int latestVersion = findLatestVersion();
+        File oldFile = recoveryLogDir.resolve(baseFileName + latestVersion + ".log").toFile();
+        if (oldFile.exists()) {
+            existingLogs = readLogsFromFile(oldFile);
+            if (deleteOldLogs) {
+                File[] files = recoveryLogDir.toFile().listFiles(
+                        (dir, name) -> name.matches(baseFileName + "(\\d+)\\.log")
+                );
+                for (File file: files) {
+                        file.delete();
+                    }
+                }
+            }
+        File newFile = recoveryLogDir.resolve(baseFileName + (latestVersion + 1) + ".log").toFile();
+        try {
+            Files.createDirectories(recoveryLogDir); // create directory if not exists
+            newFile.createNewFile();
+            initAppendChannel(newFile);
+            if (existingLogs == null) {
+                return newFile;
+            }
+            // write existing unfinished logs to the new file
+            if (!existingLogs.isEmpty()) {
```

Comment:
Do we need the condition here as we check in the `cleanUpFinishedLogs()` function also whether the `existingLogs` is empty